### PR TITLE
Mimetype list integrity check should not fail if it's changed

### LINKED ISSUE #10411: Integrity fails if mime type list is extended
* extend mimetype list
* run `occ maintenance:mimetype:update-js`
* check integrity
* expected: no issues
* actual: integrity checks don't pass

@nickvergessen @rullzer Ideas how to prevent this? Exclude the file from the check once we see that there are mimetypes added?

### PR DESCRIPTION
**Pull request by xh3n1**
**Issue**
Integrity check failed if mimetype list was changed. The expected behaviour would be that the integrity check would still be passing.

Fixes https://github.com/nextcloud/server/issues/10411

**Fix**
The fix is aiming to still perform some kind of integrity check over the mimetype list. It would still fail for manually changed mimetype lists but will pass for any valid ones.

This is done by:

Regenerating the mimetype list with all mimetype aliases
If the genereated file is same as the one in the disk, then generate the sha512 of the mimetype list file with only default aliases since they are the only ones that are signed.
I have also added "application/internet-shortcut": "link" in resources/config/mimetypealiases.dist.json 
Because it's had been manually added in the generated list https://github.com/nextcloud/server/pull/6297/files#diff-6bcc9a99f4c75054294c154bc8e2f490 but not when you generated it via maintenance:mimetype:update-js .

**Test plan**
Create a mimetypealiases.json file with custom mimetypes and save it in config dir.
Run ./occ maintenance:mimetype:update-js 
And then ./occ integrity:check-core  
With these changes it should work, without it the integrity check would fail.

PS. I have written the tests for this locally, but I having issues with running PHPUnit locally.

### CODE REVIEW

File: core/Command/Maintenance/Mimetype/GenerateMimetypeFileBuilder.php [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+class GenerateMimetypeFileBuilder
+{
+	/**
+	 * Generate mime type list file
+	 * @param $aliases
+	 * @return string
+	 */
+	public function generateFile(array $aliases): string {
+		// Remove comments
+		$keys = array_filter(array_keys($aliases), function($k) {
```
Comment (kesselb) [inline core/Command/Maintenance/Mimetype/GenerateMimetypeFileBuilder.php]: I think we can improve this a bit. It's possible to pass a flag `ARRAY_FILTER_USE_KEY` to [array_filter](https://www.php.net/manual/en/function.array-filter.php). We can omit the foreach-unset below and call the array_filter on `$aliases`.

Comment (MorrisJobke) [reply core/Command/Maintenance/Mimetype/GenerateMimetypeFileBuilder.php]: As this is code copied over from the old location it's fine to leave it as it is now in this PR and change the code in a separate PR to separate concerns properly. But still a nice finding 👍

**Comment by MorrisJobke** [comment]
> PS. I have written the tests for this locally, but I having issues with running PHPUnit locally.

Do you mind to add them here as well?

Regarding executing unit tests locally typically `./autotest.sh sqlite tests/lib/IntegrityCheck/CheckerTest.php` should work. If not then we can have a look what is not working.

File: lib/private/IntegrityCheck/Checker.php [hunk 2]
```diff
@@ -193,6 +200,14 @@ private function generateHashes(\RecursiveIteratorIterator $iterator,
 					continue;
 				}
 			}
+			if($filename === $this->environmentHelper->getServerRoot().'/core/js/mimetypelist.js') {
```
Comment (MorrisJobke) [inline lib/private/IntegrityCheck/Checker.php]:
```suggestion
			if ($filename === $this->environmentHelper->getServerRoot() . '/core/js/mimetypelist.js') {
```

**Comment by xh3n1** [comment]
@MorrisJobke Thanks, I had to downgrade my PHPUnit version to make it work and I will push my test later.

**Review by MorrisJobke — APPROVED** [review]
Beside that the code looks good and makes sense 👍

**Comment by MorrisJobke** [comment]
> @MorrisJobke Thanks, I had to downgrade my PHPUnit version to make it work and I will push my test later.

Ah ... right. We only run on version 5 or 6 and not on the latest version.

**Comment by MorrisJobke** [comment]
Let me rebase to trigger CI again here.

**Comment by MorrisJobke** [comment]
> @MorrisJobke Thanks, I had to downgrade my PHPUnit version to make it work and I will push my test later.

I just noticed that the tests are not yet included. Mind to push them? Feel free to overwrite my push, because I just rebased on master and it went through without conflicts.

**Review by skjnldsv — APPROVED** [review]

**Comment by skjnldsv** [comment]
Welcome back @xh3n1 :wave: :D

Tests failures are unrelated.

**Comment by xh3n1** [comment]
Thanks @skjnldsv :smile:
@MorrisJobke Sorry, I have pushed them now. The test was tricky and I asked for help from @LukasReschke :wink:

**Comment by skjnldsv** [comment]
Failures still seems unrelated :)

**Comment by MorrisJobke** [comment]
@xh3n1 Nice! Thanks for your help on this 👍

**Pull request merged**